### ISSUE TITLE
Meta-Templating Proposal with multiple Descriptors

### ISSUE THREAD

**Issue by christopherjbaker** [canjs/canjs#3248]
## TLDR
A generic templating system with a very simple external interface, allowing multiple definition types, as per the user's preferences, that all work interchangeably.

## Inspiration
While looking into making a templating proposal using JSX (coming soon...), it struck me at how simple the underlying API is: it is a single function call. As long as it makes that call the right way, the system doesn't know or care if you are actually using JSX.

## Details

(As it is an already established spec, I would tend towards the JSX api as the baseline for ours, adding what we need as necessary.)

Element creation is done via calls to `createElement(tagName, props, ...children)`, so a div with a class of "foo" which contains the text "hello world" would look like `createElement('div', { class: 'foo' }, 'hello world')`. If "hello" and "world" were in spans with a space between, it would look like `createElement('div', { class: 'foo' }, createElement('span', null, 'hello'), ' ', createElement('span', null, 'world'))`.

The code that actually runs the `createElement` calls is inside a function, and that function is run for every update, and its result diffed, updated, etc.

I'm not sure what the best way to communicate bindings here might be. Events could be specified by prefixing the name with `$`, similar to can-stache.

## Benefits

The main benefit here is that there can be multiple templating engines, which all are part of the same system. If the main repo is "can-template", there could be a "can-template-jsx", "can-template-stache", and ["can-template-pug"](https://github.com/canjs/canjs/issues/3197), all of which can be used within the same project. Each of those would export a "createTemplate" function, which would take whatever configs it needs (probably primarily just the template as a string), and return a function which called `createElement` when run.

One could even create special-purpose engines which take configs already created by client apis, rather than trying to make them fit into a can-stache template.

## Example

// template-engine
```javascript
const createElement = require('template').createElement;
module.exports = createTemplate() {
  return function(viewModel) {
    return createElement('div', null, viewModel.text); // this is a really useless engine, which always creates a div
  };
};
```

// app.js
```javascript
const createRenderer = require('template').createRenderer;
const createTemplate = require('template-engine');

let renderer = createRenderer(createTemplate());
let frag = renderer({ text: 'hello world' });
```
